Libraries/LibGfx/VulkanContext.cpp

```diff
@@ -47,16 +48,58 @@ static ErrorOr<VkPhysicalDevice> pick_physical_device(VkInstance instance)
     devices.resize(device_count);
     vkEnumeratePhysicalDevices(instance, &device_count, devices.data());
 
+#define __EXTENSION_SCORE(name, score) __EXTENSION_SCORE_HELPER(name, score)
+#define __EXTENSION_SCORE_HELPER(name, score) { name##sv, score }
+
+    static HashMap<StringView, int> extension_scores {
+#ifdef USE_VULKAN_IMAGES
+        __EXTENSION_SCORE(VK_KHR_EXTERNAL_MEMORY_FD_EXTENSION_NAME, 10),
+        __EXTENSION_SCORE(VK_EXT_IMAGE_DRM_FORMAT_MODIFIER_EXTENSION_NAME, 10),
+#endif
+    };
+
+#undef __EXTENSION_SCORE
+#undef __EXTENSION_SCORE_HELPER
+
     VkPhysicalDevice picked_device = VK_NULL_HANDLE;
-    // Pick discrete GPU or the first device in the list
+    int picked_score = 0;
+    VkPhysicalDeviceProperties device_properties;
+    Vector<VkExtensionProperties> extension_properties;
+    // Score devices based on their type and feature support, pick device with the highest score
     for (auto const& device : devices) {
         if (picked_device == VK_NULL_HANDLE)
             picked_device = device;
 
-        VkPhysicalDeviceProperties device_properties;
         vkGetPhysicalDeviceProperties(device, &device_properties);
+
+        int score = 0;
+
+        // Prefer discrete GPUs but also give priority to integrated GPUs
         if (device_properties.deviceType == VK_PHYSICAL_DEVICE_TYPE_DISCRETE_GPU)
+            score += 1000;
+        else if (device_properties.deviceType == VK_PHYSICAL_DEVICE_TYPE_INTEGRATED_GPU)
+            score += 100;
+
+        uint32_t extension_count = 0;
+        vkEnumerateDeviceExtensionProperties(device, nullptr, &extension_count, nullptr);
+        if (extension_count > extension_properties.size())
+            extension_properties.resize(extension_count);
```

Comment:
Instead of doing `if (extension_count > extension_properties.size())`, you can also use `extension_properties.resize_and_keep_capacity(extension_count)` to prevent any reallocation if the size is reduced for this device.